test/unit/src/utils/EntityTransformer.spec.ts

```diff
@@ -68,4 +72,92 @@ describe('entity transformer', () => {
 
     expect(normalizedEntity.id).toBe('myCustomId');
   });
+
+  it('should work for nested entities', () => {
+    const testingClass1 = new MyRootEntity();
+    testingClass1.id = '25u46fhno';
+    testingClass1.foo = 1;
+    {
+      const nestedClass = new MyNestedEntity();
+      nestedClass.name = 'name';
+      testingClass1.nestedItem = nestedClass;
+    }
+    {
+      const nestedClass = new MyNestedEntity();
+      nestedClass.name = 'name-array';
+      testingClass1.nestedItems = [nestedClass];
+    }
+
+    const normalizedEntity = entityTransformer.normalize(testingClass1);
+    expect(normalizedEntity.id).toBe('25u46fhno');
+    expect(normalizedEntity.data.foo).toBe(1);
+    expect(normalizedEntity.data.nestedItem).toEqual({ name: 'name' });
+    expect(normalizedEntity.data.nestedItems).toEqual([
+      {
+        name: 'name-array',
+      },
+    ]);
+
+    const denormalizedEntity = entityTransformer.denormalize(
+      MyRootEntity,
+      normalizedEntity,
+    );
+
+    expect(denormalizedEntity).toMatchObject(testingClass1);
+
+    const normalizedEntityRetried =
+      entityTransformer.normalize(denormalizedEntity);
+    expect(normalizedEntityRetried).toMatchObject(normalizedEntity);
+  });
+
+  it('should work for nested entities - optional fields', () => {
+    const testingClass1 = new MyRootEntity();
+    testingClass1.id = '25u46fhno';
+    testingClass1.foo = 1;
+
+    // remove block in the future
+    // see: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#the-usedefineforclassfields-flag-and-the-declare-property-modifier
+    {
+      const recentTsTarget = Object.keys(testingClass1).includes('nestedItem');
+      if (!recentTsTarget) {
+        console.debug('older TS target');
+        // make test object look like what newer TS targets generate
+        Object.assign(testingClass1, {
+          nestedItem: undefined,
+          nestedItems: undefined,
+        });
+      }
+    }
```

Comment:
This is definitely a hack. I thought about enabling the flag mentioned in those docs but it would probably be cleaner just to update the TS target. I'm not sure how extensive those changes would end up being or if you're interested in doing that. I might be able to help if you were.